Vault weapon grouping/separators

JakeSidSmith opened the pull request:
~This is a work in progress. At the bottom is a checklist of items that still need to be completed/considered.~

Fixes #7206
Fixes #3613

This PR adds ~2~ 1 option (dropdown) to the settings page, ~one~ allowing the user to define a property by which to group vault items, ~and another that allows them to tweak how the groupings are displayed~. I've limited the properties by which items can be grouped as many were not appropriate for displaying (e.g. power level and acquired date end up with far too many groups).

I've opted to use dropdown for these options as opposed to adding checkboxes or other inputs to the sort options (as mentioned in some issues), as this keep the UI a lot cleaner and doesn't involve the need to enforce only a single grouping (which would be required if the option was added to custom sort items).

~In any grouping display setting the groups will be prefixed with an icon (yet to be implemented), much like how class armor has the class icon. In the "Separated" view groups are all displayed on separate rows with horizontal lines between each, whereas the "Best Fit" view will display groups without the horizontal lines, and will include multiple groups in a single row if there is enough space.~

All groups except for rarity are prefixed with an icon, much like the class specific armor groups.

See comments for discussion and updated screenshots.

Screenshots

<img width="502" alt="Screenshot 2023-09-18 at 10 49 58" src="https://github.com/DestinyItemManager/DIM/assets/5850625/bd527d95-a3a0-41c4-9d0b-f00c64065880">
<img width="511" alt="Screenshot 2023-09-18 at 10 50 04" src="https://github.com/DestinyItemManager/DIM/assets/5850625/b79f9dae-bc90-4b5e-a9b3-51e93daea4eb">
<img width="508" alt="Screenshot 2023-09-18 at 10 50 11" src="https://github.com/DestinyItemManager/DIM/assets/5850625/c8cd782c-3d02-441c-947a-5c4154e7825a">
Grouping with "Best Fit"
<img width="1389" alt="Screenshot 2023-09-18 at 10 50 51" src="https://github.com/DestinyItemManager/DIM/assets/5850625/1527d742-b58f-4a1f-b957-7bc68c1a45d2">
Grouping with "Separated"
<img width="1381" alt="Screenshot 2023-09-18 at 10 51 08" src="https://github.com/DestinyItemManager/DIM/assets/5850625/917ad997-f30c-4c51-9fcb-0398391826da">

To consider:
- Is it worth adding any of these groupings to the armor/inventory items?
- How can this be extended to allow custom grouping (e.g. grouping by pvp/pve that are mentioned in several issues that I now cannot find)?

TODO:
- [x] ~Extract conditional grouping within StoreBucket to its own component~ - only groups weapons
- [x] Add doc comments to new utils/selectors
- [x] Add tests for new utils/selectors
- [x] ~Add icons for each grouping type (including an "unknown"/"ungrouped" icon)~ - removed icons/tiles
- [ ] Add types to DIM API types?
- [x] ~Remove grouping from armor/inventory items?~ - only groups weapons
- [x] ~Improve grouping display of armor/inventory items?~ - only groups weapons

bhollis commented:
It's a nice start! Code looks like it's going in the right direction. Some stuff to think about:

* Best fit doesn't look super good - lots of half-filled lines. Perhaps it's better to do that one the way we did armor - flatten out the lists and just insert the group separators in the grid. The downside is you can have awkward situations like the separator tile falling at the end of a row. Still, "best fit" is awkward enough right now that it definitely shouldn't be the default, and probably shouldn't be an option as is.
* For some groupings, the order of the group is somewhat arbitrary. Weapon type, which IMO is the most interesting (maybe the only interesting) grouping, uses an ordering that I probably wouldn't have if I had chosen the order by hand.
* I'm not sure this brings much value to armor. Most of the groupings don't even apply to armor. I'd remove it.
* I'm repeating myself, but everything other than weapon type and ammo type are trivially distinguishable if you simply sort by that property first (e.g. tag or rarity is very clear if it's your top sort). Putting them on different lines doesn't offer much.
* The separated view isn't bad - it probably doesn't need the line if it has a tile that explains it. But I wonder if they even need that tile - maybe it's super obvious, except for maybe weapon type and ammo type.

JakeSidSmith commented:
Thank you. 😊 

> * Best fit doesn't look super good - lots of half-filled lines. Perhaps it's better to do that one the way we did armor - flatten out the lists and just insert the group separators in the grid. The downside is you can have awkward situations like the separator tile falling at the end of a row. Still, "best fit" is awkward enough right now that it definitely shouldn't be the default, and probably shouldn't be an option as is.

I'm happy to remove the best fit view. I'm not a fan either.

> * For some groupings, the order of the group is somewhat arbitrary. Weapon type, which IMO is the most interesting (maybe the only interesting) grouping, uses an ordering that I probably wouldn't have if I had chosen the order by hand.

All the groupings are based on the existing sort options, so as to maintain consistency with the sorts. I believe weapon types are alphabetical?

> * I'm not sure this brings much value to armor. Most of the groupings don't even apply to armor. I'd remove it.

Agreed.

> * I'm repeating myself, but everything other than weapon type and ammo type are trivially distinguishable if you simply sort by that property first (e.g. tag or rarity is very clear if it's your top sort). Putting them on different lines doesn't offer much.

I agree that rarity is fairly obvious (due to the colorings), but I personally wanted to implement this feature to distinguish between the tags (failing a way to manually group items). As all the tag icons are very small and the same color it's hard to see where one ends and another begins.

> * The separated view isn't bad - it probably doesn't need the line if it has a tile that explains it. But I wonder if they even need that tile - maybe it's super obvious, except for maybe weapon type and ammo type.

The tile was basically a result of the best fit view. More code and harder to maintain if one view has icons and another didn't, or that one was flat and another nested. Happy to remove them if we're just going with the expanded view.

bhollis commented:
> All the groupings are based on the existing sort options, so as to maintain consistency with the sorts. I believe weapon types are alphabetical?

Yeah, I don't use sort by weapon type, and the grouping makes it even more obvious that it's not a particularly useful sort. I'm not sure what would be more useful though - ordering them by ammo type is just including another sort.

> As all the tag icons are very small and the same color it's hard to see where one ends and another begins.

That's interesting - I find their shapes quite distinctive, but I could see the argument.

JakeSidSmith commented:
I was actually going to suggest (as a separate issue) that the tag icons each receive a unique color, but I understand that this is more work than it sounds due to the theming.

bhollis commented:
They're also all the same color on purpose - that green is the only thing we can use that stands out against all the different item icons.

JakeSidSmith commented:
> They're also all the same color on purpose - that green is the only thing we can use that stands out against all the different item icons.

Ah, I had not considered that. 😕

JakeSidSmith commented:
I have now...

- Removed armor grouping
- Renamed the groupings as they are specific to weapons
- Removed the best fit view (and display options)
- Removed tiles/icons rendered for each group
- Added comments to grouping utils
- Added tests for grouping selectors

I'm not sure if/how the API types are updated. Please advise on how to proceed. 😊 

Updated screenshots:

<img width="533" alt="Screenshot 2023-10-16 at 13 16 43" src="https://github.com/DestinyItemManager/DIM/assets/5850625/cc42f0a0-6f56-4f22-9108-ca8553f6e730">

<img width="850" alt="Screenshot 2023-10-16 at 13 14 44" src="https://github.com/DestinyItemManager/DIM/assets/5850625/ceb59be3-b3f9-4adf-83f5-bb3fc83f1c2c">

bhollis commented:
Looks good, definitely works as advertised. A couple things I noticed:

* Without an icon, it can be hard to understand what a category contains. Even for more obvious ones like damage type it takes a second to figure out. I think we really should have them, but it makes it hard to support as many options.
* This says it only affects weapons, but it clearly also groups my armor (at least in single character mode) - e.g. when sorting by type, FoTL masks are separated, when sorting by rarity exotics are separate. For some of the sorts (rarity, tag) that's actually nice, for some it doesn't mean much.
* We probably need to have custom / handcrafted orders for most of these. For example, sorting by tag puts my "archived" items first.
* I can sorta see why you'd choose most of these, but "Deepsight" doesn't have a lot of value.

API types are updated here: https://github.com/DestinyItemManager/dim-api/blob/master/api/shapes/settings.ts - it's OK to just update `initial-settings`, I move new settings to the API types after a while.

JakeSidSmith commented:
I cannot update my branch from master due to a prettier error in `src/index.html`.

Tried to fix this myself, but seems prettier just doesn't like the lodash template syntax.

<img width="914" alt="Screenshot 2023-11-01 at 10 43 54" src="https://github.com/DestinyItemManager/DIM/assets/5850625/b90ee7e2-3fca-44d3-a451-336bffc80986">

bhollis commented:
I updated it for you. Not sure what you were running into, we removed prettier running on HTML in a recent commit on `master`.

JakeSidSmith commented:
Thank you. Not sure why, but after merging master into my branch and resolving conflicts the husky pre-commit ran and complained about the `src/index.html`.

I'm going to reduce the number of things we can group by to only those with 3 or more possible groups.

<img width="273" alt="Screenshot 2023-11-03 at 14 12 27" src="https://github.com/DestinyItemManager/DIM/assets/5850625/7de20f3a-feee-46ef-a629-4a78731f62d4">

Removing

<img width="277" alt="Screenshot 2023-11-03 at 14 12 37" src="https://github.com/DestinyItemManager/DIM/assets/5850625/c72608a5-ed68-454b-9d3d-74b1fb657798">

Will also have a look at ensuring this only groups weapons - I never use the single character view, so hadn't noticed the accidental armor grouping.

And then have a look into adding some icons.

JakeSidSmith commented:
Oh, and improve the tag grouping. I'm guessing this should match the order displayed in brackets on the option:
- Favorite
- Keep
- Junk
- Infuse
- Archive

JakeSidSmith commented:
Tag ordering has been improved, and have added icons for all groupings except rarity (as the colors are fairly obvious and I'm not sure there's a suitable icon for these).

Think I'm happy for this to ship now. 😁 Thoughts?

<img width="853" alt="Screenshot 2023-11-07 at 17 33 51" src="https://github.com/DestinyItemManager/DIM/assets/5850625/0ef15eb8-9b4d-452f-b251-c4e0a93bb5bc">
<img width="855" alt="Screenshot 2023-11-07 at 17 34 11" src="https://github.com/DestinyItemManager/DIM/assets/5850625/b2b8689c-8997-4211-a24a-04899d5e3286">
<img width="841" alt="Screenshot 2023-11-07 at 17 34 29" src="https://github.com/DestinyItemManager/DIM/assets/5850625/231404df-613f-46e5-93a9-35ac58e80478">
<img width="523" alt="Screenshot 2023-11-07 at 17 34 56" src="https://github.com/DestinyItemManager/DIM/assets/5850625/ad04bc49-d53a-4f2a-8f9d-e176160e5ddc">

bhollis commented:
We're so close, I can taste it! I love the way this looks and works now. A couple last bits:

1. The ammo icons are a touch small compared to the others.
2. In single character mode, no armor shows up at all.

bhollis commented:
That's my bad, I was testing a different character who doesn't actually have any items in the vault 😅

bhollis commented on src/app/shell/item-comparators.ts:
```diff
@@ -109,6 +110,114 @@ const TAG_ITEM_COMPARATORS: {
 
+export type VaultGroupValue = string | number | boolean | undefined;
+
+interface MutableVaultGroup {
+  groupingValue: VaultGroupValue;
+  iconValue?: VaultGroupIconValue;
+  items: DimItem[];
+}
+
+interface VaultGroup {
+  groupingValue: VaultGroupValue;
+  iconValue?: VaultGroupIconValue;
+  items: readonly DimItem[];
+}
```
There isn't really a need to enforce immutability via the type here, but if you want to, there's also `Draft<VaultGroup>` from the `immer` types.

bhollis commented on src/app/shell/item-comparators.ts:
```diff
@@ -109,6 +110,114 @@ const TAG_ITEM_COMPARATORS: {
+  },
+};
+
+const TAG_ORDER = ['favorite', 'keep', 'junk', 'infuse', 'archive'];
```
`tagConfig` in `dim-item-info.ts` has the preferred sort order of tags.

bhollis commented on src/app/shell/item-comparators.ts:
```diff
@@ -240,3 +349,69 @@ export function sortItems(
+  vaultGrouping: string,
+  getTag: (item: DimItem) => TagValue | undefined,
+): readonly VaultGroup[] {
+  const groupingValueGetter = GROUPING_VALUE_GETTERS[vaultGrouping];
+  const iconValueGetter = ICON_VALUE_GETTERS[vaultGrouping];
+  const comparator = GROUP_BY_COMPARATORS[vaultGrouping];
```
Rather than have three parallel maps, you could have one map where each entry is an object consisting of value, icon, and comparator

JakeSidSmith replied:
Yeah, will do - saves on some logic.

bhollis commented on src/app/shell/item-comparators.ts:
```diff
@@ -240,3 +349,69 @@ export function sortItems(
+
+    const groupingValue = groupingValueGetter(item, getTag);
+
+    if (typeof groupingValue === 'undefined') {
```
It seems sufficient to do this, like we do elsewhere:
```suggestion
    if (groupingValue === undefined) {
```

bhollis commented on src/app/shell/item-comparators.ts:
```diff
@@ -240,3 +349,69 @@ export function sortItems(
+    const indexOfUngrouped = grouped.findIndex((g) => g.groupingValue === undefined);
+
+    if (!groupingValueGetter) {
+      // If there is not an ungrouped group, create one
```
This is an interesting insertion grouping approach, but it might be easier and more efficient to just collect things in a Map?

JakeSidSmith replied:
I'm not sure what you mean/how that would work. Can you expand?

bhollis replied:
e.g.:

```ts
const groupedItems = Map.groupBy(items, (i) => groupingValueGetter(item, getTag));
return Array.from(groupedItems.entries(), ([groupingValue, items]) => ({ groupingValue, items, iconValue: iconValueGetter?.(items[0], getTag) })).sort(comparator);
```

bhollis commented on src/app/inventory-page/StoreBucket.tsx:
```diff
@@ -72,19 +80,25 @@ const StoreBucketInner = memo(function StoreBucketInner({
   const equippedItem = isVault ? undefined : items.find((i) => i.equipped);
-  const unequippedItems = isVault ? sortItems(items) : sortItems(items.filter((i) => !i.equipped));
+  const unequippedItems =
+    isVault && bucket.type && WEAPON_BUCKET_TYPES.includes(bucket.type)
```
Can we not do this?
```suggestion
    isVault && bucket.inWeapons
```

JakeSidSmith replied:
Didn't know that existed. 😁 Thanks.

bhollis commented on src/app/inventory-page/StoreBucket.tsx:
```diff
@@ -108,15 +122,31 @@ const StoreBucketInner = memo(function StoreBucketInner({
         </StoreBucketDropTarget>
       )}
       <StoreBucketDropTarget
+        grouped={isVault && Boolean(vaultWeaponGroupingSetting)}
```
If you're only using this as a boolean you should probably select a boolean, e.g. `vaultWeaponGroupingEnabledSelector`

JakeSidSmith replied:
The actual value is used elsewhere, I just need to know if it's truthy in this case.

bhollis replied:
Right - it's a nitpick, but `useSelector` will trigger a re-render whenever the output value of the selector changes - you have only two useful values (true/false) but your output can be one of many values. It doesn't really matter in practice, it's just good habit.

JakeSidSmith replied:
Oh, I see what you mean. I can add another selector. 😊

bhollis commented on src/app/dim-ui/WeaponGroupingIcon.tsx:
```diff
@@ -0,0 +1,104 @@
+  iconValue: VaultGroupIconValue;
+  className?: string;
+}) {
+  const vaultWeaponGroupingSetting = useSelector(vaultWeaponGroupingSettingSelector);
+
```
Is this necessary? I don't think you're using this component if the setting is disabled.

JakeSidSmith replied:
The alternative is to pass the grouping setting in. I thought this was cleaner, but don't mind either way.

bhollis replied:
I don't think you need it at all if you pass in the `iconValue` as a tagged union.

JakeSidSmith replied:
Yeah, will remove when I add that tagged union. 😊

bhollis commented on src/app/shell/item-comparators.ts:
```diff
@@ -109,6 +110,114 @@ const TAG_ITEM_COMPARATORS: {
+  | DestinyDamageTypeDefinition
+  | DestinyAmmunitionType
+  | number[]
+  | undefined
+  | null;
```
It would be good to avoid `null` here in favor of `undefined` - handling both makes things unnecessarily complicated.

JakeSidSmith replied:
I opted for `null` in addition to `undefined` so that items that explicitly don't have an icon have `null`, whereas items that should have an icon but it is missing result in `undefined`. Just makes debugging a little easier.

bhollis replied:
I generally try hard to avoid encoding semantic meaning in the difference between `null`/`undefined`

JakeSidSmith replied:
I guess this will also be removed when I move to tagged union.

bhollis commented on src/app/dim-ui/WeaponGroupingIcon.tsx:
```diff
@@ -0,0 +1,104 @@
+
+  if (
+    vaultWeaponGroupingSetting === 'typeName' &&
+    typeof iconValue === 'object' &&
+    // eslint-disable-next-line no-implicit-coercion
+    !!iconValue &&
+    !('hash' in iconValue)
```
It can help to write some type predicates for things like this:

```
function isCategoryHashes(iconValue: VaultGroupIconValue): iconValue is number[]) {
  ...
}
```

That said, this is pretty complicated. What if `VaultGroupIconValue` was just a tagged union?:

```
type VaultGroupIconValue = { type: 'itemCategoryHash'; value: number[] } | ...;
```

Then it's super easy to just switch on `iconValue.type` and have all your cases under that.

JakeSidSmith replied:
I'd actually considered that approach, so if you're onboard I'll refactor it to work this way. 😊

bhollis commented on src/app/dim-ui/WeaponGroupingIcon.tsx:
```diff
@@ -0,0 +1,104 @@
+  if (
+    vaultWeaponGroupingSetting === 'typeName' &&
+    typeof iconValue === 'object' &&
+    // eslint-disable-next-line no-implicit-coercion
+    !!iconValue &&
```
```suggestion
    Boolean(iconValue) &&
```

JakeSidSmith replied:
This is a type-guard. `Boolean()` doesn't prove that the thing that's passed in exists, whereas `!!` does.

The alternative is to manually create a type guard for this, but it's less type-safe.

JakeSidSmith replied:
I actually used to enforce `Boolean` with linting myself until I got more into TypeScript and then dropped it in favor of double negation.

bhollis replied:
Yeah, that's https://github.com/microsoft/TypeScript/issues/16655. I don't love how implicit/magical `!!` is but I'll consider reversing the lint. Generally we try to structure types such that this isn't an issue, since it's confusing to have to do this type of thing.

bhollis replied:
Also see https://github.com/typescript-eslint/typescript-eslint/issues/6316#issuecomment-1376059324

JakeSidSmith commented:
@bhollis have made all suggested changes. 😁

bhollis reviewed (approved):
Great! Let's get this merged!
